chronos/apps/http/multipart.nim

```diff
@@ -0,0 +1,471 @@
+      stream*: HttpBodyReader
+    of MultiPartSource.Buffer:
+      discard
+    firstTime: bool
+    buffer: seq[byte]
+    offset: int
+    boundary: seq[byte]
+    counter: int
+
+  MultiPartReaderRef* = ref MultiPartReader
+
+  MultiPart* = object
+    case kind: MultiPartSource
+    of MultiPartSource.Stream:
+      breader: HttpBodyReader
+      stream: BoundedStreamReader
+    of MultiPartSource.Buffer:
+      discard
+    buffer: seq[byte]
+    headers: HttpTable
+    counter: int
+    name*: string
+    filename*: string
+
+  MultipartError* = object of HttpCriticalError
+  MultipartEOMError* = object of MultipartError
+
+  BChar* = byte | char
+
+proc startsWith(s, prefix: openarray[byte]): bool {.
+     raises: [Defect].} =
+  var i = 0
+  while true:
+    if i >= len(prefix): return true
+    if i >= len(s) or s[i] != prefix[i]: return false
+    inc(i)
+
+proc parseUntil(s, until: openarray[byte]): int {.
+     raises: [Defect].} =
+  var i = 0
+  while i < len(s):
+    if len(until) > 0 and s[i] == until[0]:
+      var u = 1
+      while i + u < len(s) and u < len(until) and s[i + u] == until[u]:
+        inc u
+      if u >= len(until): return i
+    inc(i)
+  -1
+
+func setPartNames(part: var MultiPart): HttpResult[void] {.
+     raises: [Defect].} =
+  if part.headers.count("content-disposition") != 1:
+    return err("Content-Disposition header is incorrect")
+  var header = part.headers.getString("content-disposition")
+  let disp = parseDisposition(header, false)
+  if disp.failed():
+    return err("Content-Disposition header value is incorrect")
+  let dtype = disp.dispositionType(header.toOpenArrayByte(0, len(header) - 1))
+  if dtype.toLowerAscii() != "form-data":
+    return err("Content-Disposition type is incorrect")
+  for k, v in disp.fields(header.toOpenArrayByte(0, len(header) - 1)):
+    case k.toLowerAscii()
+    of "name":
+      part.name = v
+    of "filename":
+      part.filename = v
+    else:
+      discard
+  if len(part.name) == 0:
+    part.name = $part.counter
+  ok()
+
+proc init*[A: BChar, B: BChar](mpt: typedesc[MultiPartReader],
+                               buffer: openarray[A],
+                               boundary: openarray[B]): MultiPartReader {.
+     raises: [Defect].} =
+  ## Create new MultiPartReader instance with `buffer` interface.
+  ##
+  ## ``buffer`` - is buffer which will be used to read data.
+  ## ``boundary`` - is multipart boundary, this value must not be empty.
+  doAssert(len(boundary) > 0)
+  # Our internal boundary has format `<CR><LF><-><-><boundary>`, so we can
+  # reuse different parts of this sequence for processing.
+  var fboundary = newSeq[byte](len(boundary) + 4)
+  fboundary[0] = 0x0D'u8
+  fboundary[1] = 0x0A'u8
+  fboundary[2] = byte('-')
+  fboundary[3] = byte('-')
+  copyMem(addr fboundary[4], unsafeAddr boundary[0], len(boundary))
+  # Make copy of buffer, because all the returned parts depending on it.
+  var buf = newSeq[byte](len(buffer))
```

Comment:
Its not working for procedures which are able to accept not only `byte` but also `char`. Because @buffer where buffer is `openarray[char]` will become `seq[char]` but for further processing i need to have single buffer and it should be `seq[byte]`.